### PR TITLE
Added support for namespaces during client initialization

### PR DESCRIPTION
**Pull request by stenio123**
Update to support new Namespaces feature of Vault 0.11.0.

Full sys/namespace support not included.

### CODE REVIEW

**Comment by stenio123** [comment]
Example on how to validate can be found here:
https://gist.github.com/stenio123/30c6a444576b7d4401c473a9ed4aeaa9

**Comment by chrishoffman** [comment]
One question I have here is should this be a parameter of the client or should you be able to pass it per request.  Both have their advantages but for some workflows, creating a new client seems like a heavy way for a API client to act.

File: lib/vault/defaults.rb [hunk 1]
```diff
@@ -61,6 +65,13 @@ def token
         nil
       end
 
+
+      # Vault Namespacem if any.
```
Comment (kalafut) [inline lib/vault/defaults.rb]: s/Namespacem/Namespace

File: spec/unit/defaults_spec.rb [hunk 2]
```diff
@@ -57,6 +57,20 @@ module Vault
...
 
+    describe ".namespace" do
+      it "uses ENV['VAULT_NAMESPACE'] if present" do
+        with_stubbed_env("VAULT_NAMESPACE" => "test") do
+          expect(Defaults.namespace).to eq("test")
+        end
+      end
+
+      it "falls back to the default VAULT_NAMESPACE" do
+        with_stubbed_env("VAULT_ADDR" => nil) do
```
Comment (kalafut) [inline spec/unit/defaults_spec.rb]: Should be "VAULT_NAMESPACE", right?

File: lib/vault/defaults.rb [hunk 3]
```diff
@@ -11,6 +11,10 @@ module Defaults
     # @return [String]
     VAULT_DISK_TOKEN = Pathname.new("#{ENV["HOME"]}/.vault-token").expand_path.freeze
 
+    # Namespace from env variable
+    # @return [String]
+    VAULT_NAMESPACE = ENV["VAULT_NAMESPACE"]
```
Comment (kalafut) [inline lib/vault/defaults.rb]: Is this needed?

File: spec/unit/defaults_spec.rb [hunk 4]
```diff
@@ -57,6 +57,20 @@ module Vault
       end
     end
 
+    describe ".namespace" do
```
Comment (kalafut) [inline spec/unit/defaults_spec.rb]: Isn't the desired behavior to use the env var if present, and to be nil if not, such that the header isn't added at all?

**Comment by pablomolnar** [comment]
I had to monkey patch the lack of namespace support with the following code:

```ruby
DEFAULT_HEADERS = Vault::Client.class_eval('DEFAULT_HEADERS').dup.merge({"X-Vault-Namespace" => "my-namespace"})
Vault::Client.const_set('DEFAULT_HEADERS', DEFAULT_HEADERS)
```

Hoping namespace feature gets added soon...

**Comment by mchugh19** [comment]
Hey all. Is there any hope of getting namespace support into the ruby lib? Once this is released, we'd still need to adjust our chef resources in order to take advantage, so this is a blocking issue.

**Comment by eheydrick** [comment]
Would love to see this merged in. Looks like it needs a rebase at least.

**Comment by isukapalli** [comment]
Is there any timeline when this gets merged and add support to Namespace? Thanks

**Comment by Valarissa** [comment]
Hey all~! I'm new to HashiCorp, but for my first real assignment, I'm giving the Ruby Client some much needed TLC. I'll be starting by looking to merge this with some additional tests and other support for Namespaces like the ability to create Clients with different namespaces!

Stay tuned!

**Comment by Valarissa** [comment]
Superseded by #229. All of @stenio123's changes are there and then some!

**Pull request closed without merging**